[Spring 경로 조회 - 2단계] 스컬(김윤기) 미션 제출합니다.

skullkim opened the pull request:
닉~ 2단계도 잘부탁해요😁

skullkim committed:
fix(PathResponse.java): 필드명 수정

API 문서와 이름을 일치시키기 위해 필드명 수정

BREAKING CHANGE: stationResponse를 stations로 변경
as-is:
public class PathResponse {

    private List<StationResponse> stationResponses;
    ...

    ...

    public PathResponse(final List<Station> stations, final int distance, final int fare) {
        this.stationResponses = stations.stream()
              ...
    }

    public List<StationResponse> getStationResponses() {
        return stationResponses;
    }

    ...
}

to-be:
public class PathResponse {

    private List<StationResponse> stations;
    ...

    ...

    public PathResponse(final List<Station> stations, final int distance, final int fare) {
        this.stations = stations.stream()
            ...
    }

    public List<StationResponse> getStations() {
        return stations;
    }

    ...
}

skullkim committed:
docs(.gitignore): 커밋 하지 않을 파일 추가

IntelliJ에서 자동 새성되는 .DS_store를 커밋에 올리지 않기 위해
.gitignore에 추가

skullkim committed:
refactor(PathController.java): 필요 없는 어노테이션 추가

생략 가능한 어노테이션 삭제

BREAKING CHANGE: 쿼리를 객체로 받기 위해 사용한 @ModelAttribute 제거
as-is:
@RestController
public class PathController {
    ...

    @GetMapping("/paths")
    public ResponseEntity<PathResponse> showPath(@ModelAttribute final PathRequest pathRequest) {
        ...
    }
}

to-be:
@RestController
public class PathController {
    ...

    @GetMapping("/paths")
    public ResponseEntity<PathResponse> showPath(final PathRequest pathRequest) {
        ...
    }
}

skullkim committed:
test(StationController.java): stationController 테스트 추가

skullkim committed:
test(LineController.java): 라인 컨트롤러 테스트 추가

라인에 대한 요청에 포함되는 데이터 유효 여부와 응답을 검증하는 테스트
추가

skullkim committed:
test(PathControllerTest.java): 요청 쿼리 유효성 검증 테스트 추가

유효하지 않은 데이터가 왔을 경우를 테스트 하는 테스트 추가

skullkim committed:
test(SectionControllerTest.java): 섹션 컨트롤러 테스트 추가

섹션 관련 요청에 포함된 데이터 유효 여부와 그에 따른 응답에 대한 테스트
추가

skullkim committed:
feat(ControllerAdvice.class): 관리할 에러 추가

spring bean validation을 사용한 검증에 따른 관리할 에러 추가

skullkim committed:
refactor(PathCalculator.java): 반환 타입 변경

여러 메서드로 나눠져 있던 값을 하나의 타임으로 묶어서 반환하게 변경

BREAKING CAHNGE: 최단거리와 최단 경로를 반환하는 두 개의 메서드를 하나로
통일
as-is:
public class PathCalculator {
    ...

    public List<Station> findShortestPath(final Station source, final Station target) {
        return dijkstraShortestPath.getPath(source, target).getVertexList();
    }

    public int findShortestDistance(final Station source, final Station target) {
        final double shortestDistance = dijkstraShortestPath.getPath(source, target).getWeight();
        return (int) shortestDistance;
    }
}

to-be:
public class PathCalculator {
    ...

    public GraphPath<Station, DefaultWeightedEdge> findShortestPath(final Station source, final Station target) {
        return dijkstraShortestPath.getPath(source, target);
    }
}

skullkim committed:
feat(schema.sql): 라인 테이블에 컬럼 추가

라인 테이블에 라인별 추가 요금을 저장하는 컬럼 추가

skullkim committed:
feat(LineDao.java): 라인 추가 요금 저장 기능 추가

레코드를 추가할 때 해당 라인이 가지는 추가 요금도 저장하는 기능 추가

skullkim committed:
feat(subway/): 라인 추가 요금 저장 기능 완료

라인 추가 요금을 요청으로 부터 저장하는 기능 완료

skullkim committed:
feat(Fare.java): 노선 추가 요금 까지 계산하는 기능 완료

지하철 요금 계산 시 노선에 따른 추가 요금도 고려하는 기능 완료

skullkim committed:
feat(discountpolicy/): 나이 할인 정책 기능 추가

나이대에 따라 금액을 할인하는 기능 완료

skullkim committed:
refactor(Fare.java): 코드 가독성 개선

메서드 분리, 조건문 조건 수정을 통해 코드 가독성 개선

BREAKING CHANGE: 조건문의 조건을 메서드로 분리, 조건문 조건 수정
as-is:
public class Fare {

    ...

    public int calculate(final int distance, final List<Station> stations, final List<Line> lines) {
        ...
        if (distance >= ADDITIONAL_DISTANCE_PER_5KM && distance < ADDITIONAL_DISTANCE_PER_8KM) {
            ...
        }
        if (distance >= ADDITIONAL_DISTANCE_PER_8KM) {
            ...
        }
        ...
    }

    ...

    private int findExtraLineFare(final Line line, final List<Station> stations) {
        final boolean doesLineContainStation = stations.stream()
                .anyMatch(line::containStation);
        if (doesLineContainStation) {
            ...
        }
        ....
    }

    ...
}

to-be:
public class Fare {

    ...

    public int calculate(final int distance, final List<Station> stations, final List<Line> lines) {
        ...
        if (ADDITIONAL_DISTANCE_PER_5KM <= distance && distance < ADDITIONAL_DISTANCE_PER_8KM) {
            ...
        }
        if (ADDITIONAL_DISTANCE_PER_8KM <= distance) {
            ...
        }
        ...
    }

    ...

    private int findExtraLineFare(final Line line, final List<Station> stations) {
        if (doesLineContainStation(line, stations)) {
            ...
        }
        ....
    }

    private boolean doesLineContainStation(final Line line, final List<Station> stations) {
        return stations.stream().anyMatch(line::containStation);
    }

    ...
}

skullkim committed:
chore(/main): 코드 스타일 수정

코딩 컨벤션을 맞추기 위해 코드 스타일 수정

skullkim committed:
refactor(Fare.java): 메서드 간략화

경로에 존재하는 라인 중 최대 경로 추가 비용을 찾는 로직 간략화

BREAKING CHANGE: calculateExtraLineFare() 로직 간략화
as-is:
public class Fare {
    ...

    private int calculateExtraLineFare(final List<Station> stations, final List<Line> lines) {
        int maxExtraLineFare = DEFAULT_EXTRA_LINE_FARE;
        for (Line line : lines) {
            maxExtraLineFare = Math.max(maxExtraLineFare, findExtraLineFare(line, stations));
        }
        return maxExtraLineFare;
    }

    private int findExtraLineFare(final Line line, final List<Station> stations) {
        if (doesLineContainStation(line, stations)) {
            return line.getExtraFare();
        }
        return DEFAULT_EXTRA_LINE_FARE;
    }

    ...
}

to-be:
public class Fare {
    ...

     private int calculateExtraLineFare(final List<Station> stations, final List<Line> lines) {
        return lines.stream()
                .filter(line -> doesLineContainStation(line, stations))
                .mapToInt(Line::getExtraFare)
                .max()
                .orElse(DEFAULT_EXTRA_LINE_FARE);
    }

    ...
}

wbluke commented on src/test/java/wooteco/subway/ui/PathControllerTest.java:
```diff
@@ -23,13 +24,22 @@ class PathControllerTest {
+        @DisplayName("입력한 데이터 중 하나라도 유효하지 않으면 지하철 경로를 조회하지 못한다.")
+        @ParameterizedTest(name = "{index} {displayName} source={0} target={1} age={2}")
+        @CsvSource(value = {"0, 1, 1", "1, 0, 1,", "1, 1, 0"})
+        void failedToShowPath(final Long source, final Long target, final int age) throws Exception {
```
Controller 단위테스트 잘 진행해 주셨네요 👍 
가능하면 에러 시 반환하는 메시지도 같이 검증해보면 좋을 것 같아요 :)

skullkim replied:
추가할게요~

wbluke commented on src/main/java/wooteco/subway/ui/PathController.java:
```diff
@@ -18,8 +18,9 @@ public PathController(final PathService pathService) {
-    public ResponseEntity<PathResponse> showPath(@ModelAttribute final PathRequest pathRequest) {
-        final PathResponse pathResponse = pathService.findShortestPath(pathRequest.getSource(), pathRequest.getTarget());
+    public ResponseEntity<PathResponse> showPath(@Valid final PathRequest pathRequest) {
+        final PathResponse pathResponse = pathService.findShortestPath(pathRequest.getSource(),
+                pathRequest.getTarget(), pathRequest.getAge());
```
Request에 필드가 많아질수록 메서드 파라미터도 같이 늘어나야 할 것 같은데요.
Service 레이어 쪽에 DTO가 있고 Request 끼리 변환해서 넘겨주는 방식은 어떨까요?
```java
service.findShortestPath(request.toServiceRequest());
```

skullkim replied:
그 방법이 훨씬 좋겠네요. 수정할게요😊

wbluke commented on src/main/java/wooteco/subway/domain/discountpolicy/AgeDiscountFactory.java:
```diff
@@ -0,0 +1,25 @@
+
+    private static final Map<Age, Supplier<AgeDiscountPolicy>> cache;
+
+    static {
+        cache = new EnumMap<>(Age.class);
+        cache.put(Age.PRESCHOOLER, PreschoolerPolicy::new);
+        cache.put(Age.CHILD, ChildDiscountPolicy::new);
+        cache.put(Age.TEEN, TeenDiscountPolicy::new);
+        cache.put(Age.ADULT, AdultDiscountPolicy::new);
+    }
```
좋네요 👍 
음 나이 구간과 정책이 추가되면 여기도 계속 추가해야 할 것 같은데 Age와 Policy를 묶은 Enum을 만들고 이를 전체 순회하면서 cache map을 만들도록 하면 어떨까요? :)

skullkim replied:
그럼 Age가 AgePolicy의 하위 클래스들의 생성자를 갖고 있도록 하면 되겠네요. 수정할게요

wbluke commented on src/main/java/wooteco/subway/domain/discountpolicy/Age.java:
```diff
@@ -0,0 +1,27 @@
+    CHILD(6, 12),
+    TEEN(13, 18),
+    ADULT(19, Integer.MAX_VALUE)
+    ;
```
Age는 나이 숫자에 대한 명명같은데 실제로는 **나이 구간**에 대한 내용이네요 :)

skullkim replied:
AgeRange가 더 명확할거 같네요. 수정하겠습니다

wbluke commented on src/main/java/wooteco/subway/domain/discountpolicy/AgeDiscountFactory.java:
```diff
@@ -0,0 +1,25 @@
+        cache.put(Age.ADULT, AdultDiscountPolicy::new);
+    }
+
+    public static AgeDiscountPolicy from(final int age) {
```
int보다 아예 Age를 받으면 어떨까요? :)

skullkim replied:
캐싱의 키를 Age로 하고 있고 그 Age를 찾는 거는 해당 팩토리의 책임이 아닐 수도 있다는 생각이 드네요. 외부에서 Age를 주입하는 방식으로 바꿀게요.

wbluke commented on src/main/java/wooteco/subway/domain/Fare.java:
```diff
@@ -1,25 +1,48 @@
+        return ageDiscountPolicy.discount(fare);
+    }
+
+    private int calculateExtraLineFare(final List<Station> stations, final List<Line> lines) {
+        return lines.stream()
+                .filter(line -> doesLineContainStation(line, stations))
+                .mapToInt(Line::getExtraFare)
+                .max()
+                .orElse(DEFAULT_EXTRA_LINE_FARE);
+    }
```
Lines 일급컬렉션을 만들어서 이 가공 기능을 담당해보면 어떨까요? :)

skullkim replied:
line 일급 컬렉션의 책임을 볼 수 도 있겠네요. 거기에 일급 컬랙션을 사용하면 파라미터가 출어들어 가독성도 오라가겠네요. 수정할게요~.

wbluke commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -21,15 +24,16 @@ public PathService(final StationRepository stationRepository, final LineReposito
-        final List<Station> stations = pathCalculator.findShortestPath(source, target);
-        final int distance = pathCalculator.findShortestDistance(source, target);
-        final int fare = (new Fare()).calculate(distance);
+        final GraphPath<Station, DefaultWeightedEdge> path = pathCalculator.findShortestPath(source, target);
+        final List<Station> stations = path.getVertexList();
+        final int distance = (int) path.getWeight();
```
GraphPath가 너무 naive한 느낌이 있어서, 이를 감싸서 추상화한 객체를 만들어보는건 어떨까요? :)

skullkim replied:
그래프 관련 라이브러리들을 서비스와 다른 레이어에서 사용하게 되면 변경에 취약한 문제가 발생하네요. GraphPath 뿐만 아니라 PathCalculator도 내부에서 DijkstraShortestPath를 사용하고 있어 도메인으로 부터 분리시키는게 좋겠네요, 인터페이스를 domain과 service에 넣고 구현체들을 infrastructure 레이어를 만들어 넣을게요

wbluke reviewed (changes requested):
2단계도 깔끔하게 잘 구현해 주셨네요 ㅎㅎ
몇 가지 수정해보면 좋을 부분들 코멘트 남겼으니 확인 부탁드려요 🙂

skullkim committed:
test(PathControllerTest.java): 에러 메시지 검증 추가

올바른 응답 에러 메시지를 확인하는 과정 추가

skullkim committed:
test(ui/): 에러 메시지 검증 로직 추가

spring bean validation을 통해 에러가 발생했을 때의 에러 메시지를
검증하는 로직 추가

skullkim committed:
refactor(PathDto.java): DTO 추가

컨트롤러에서 받은 경로 관련 응답 데이터를 서비스에서 사용하기 위한 DTO
사용

BREAKING CHANGE: PathService에 PathDto 추가
as-is:
@RestController
@Validated
public class PathController {
    ...

    @GetMapping("/paths")
    public ResponseEntity<PathResponse> showPath(@Valid final PathRequest pathRequest) {
        final PathResponse pathResponse = pathService.findShortestPath(pathRequest.getSource(),
                pathRequest.getTarget(), pathRequest.getAge());
        ...
    }
}

@Service
public class PathService {
    ...

    public PathResponse findShortestPath(final Long sourceId, final Long targetId, final int age) {
        final Station source = stationRepository.findById(sourceId);
        final Station target = stationRepository.findById(targetId);
        ...

        ...
        final int fare = (new Fare(AgeDiscountFactory.from(age))).calculate(distance, stations, lines);

        ...
    }
}

to-be:
@RestController
@Validated
public class PathController {

    ...

    @GetMapping("/paths")
    public ResponseEntity<PathResponse> showPath(@Valid final PathRequest pathRequest) {
        final PathResponse pathResponse = pathService.findShortestPath(pathRequest);
        ...
    }
}

to-be:
@Service
public class PathService {

    ...

    public PathResponse findShortestPath(final PathRequest request) {
        final PathDto pathDto = PathDto.from(request);
        final Station source = stationRepository.findById(pathDto.getSource());
        final Station target = stationRepository.findById(pathDto.getTarget());
        ...

        ...
        final int fare = (new Fare(AgeDiscountFactory.from(pathDto.getAge()))).calculate(distance, stations, lines);

        ...
    }
}

skullkim committed:
refactor(AgeDiscountFactory.java): 캐싱 로직 변경

알맞은 할인 정책을 캐싱하는 로직 변경

BREAKING CAHNGE: Age에서 캐싱을 생성하고 반환하는 방식으로 변경
as-is:
public class AgeDiscountFactory {

    private static final Map<Age, Supplier<AgeDiscountPolicy>> cache;

    static {
        cache = new EnumMap<>(Age.class);
        cache.put(Age.PRESCHOOLER, PreschoolerPolicy::new);
        cache.put(Age.CHILD, ChildDiscountPolicy::new);
        cache.put(Age.TEEN, TeenDiscountPolicy::new);
        cache.put(Age.ADULT, AdultDiscountPolicy::new);
    }

    ...
}

to-be:
public enum Age {

    PRESCHOOLER(1, 5, PreschoolerPolicy::new),
    CHILD(6, 12, ChildDiscountPolicy::new),
    TEEN(13, 18, TeenDiscountPolicy::new),
    ADULT(19, Integer.MAX_VALUE, AdultDiscountPolicy::new)
    ;

    ...
    private final Supplier<AgeDiscountPolicy> policySupplier;

    Age(final int minRange, final int maxRage, final Supplier<AgeDiscountPolicy> policySupplier) {
        ...
        this.policySupplier = policySupplier;
    }

    ...

    public static Map<Age, Supplier<AgeDiscountPolicy>> generateCache() {
        return Arrays.stream(values())
                .map(age -> Map.entry(age, age.policySupplier))
                .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
    }
}

public class AgeDiscountFactory {

    private static final Map<Age, Supplier<AgeDiscountPolicy>> cache = Age.generateCache();

    ...
}

skullkim committed:
refactor(Age.java): 클래스 이름 수정

의도에 맞지 않는 클래스 이름을 명확성을 위해 수정

BREAKING CHANGE: 클래스 이름을 Age에서 AgeRange로 수정

skullkim committed:
refactor(AgeDiscountFactory.java): 파라미터 변경

더 명확한 표현을 위해 나이를 받는 방식에서 나이에 대한 상수를 받는 방식으로 변경

BREAKING CHANGE: from 정적 메서드의 파라미터를 int에서 AgeRange로 변경
as-is:
public class AgeDiscountFactory {

    ...

    public static AgeDiscountPolicy from(final int age) {
        final Supplier<AgeDiscountPolicy> supplier = Optional.ofNullable(cache.get(AgeRange.from(age)))
        ...
        ...
    }
}

to-be:
public class AgeDiscountFactory {

    ...

    public static AgeDiscountPolicy from(final AgeRange ageRange) {
        final Supplier<AgeDiscountPolicy> supplier = Optional.ofNullable(cache.get(ageRange))
                ...
        ...
    }
}

skullkim committed:
refactor(PathCalculator.java): 라이브러리 분리

도메인에서 라이브러리를 격리시키기 위해 인터페이스로 변경
BREAKING CHANGE: PathCalculator를 인터페이스로 변경
as-is:
public class PathCalculator {

    private final DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath;

    public PathCalculator(final List<Line> lines) {
        final WeightedMultigraph<Station, DefaultWeightedEdge> graph = new WeightedMultigraph<>(
                DefaultWeightedEdge.class);
        for (Line line : lines) {
            final List<Section> sections = line.getSections();
            addSectionInGraph(graph, sections);
        }
        dijkstraShortestPath = new DijkstraShortestPath<>(graph);
    }

    private void addSectionInGraph(final WeightedMultigraph<Station, DefaultWeightedEdge> graph,
                                   final List<Section> sections) {
        for (Section section : sections) {
            final Station upStation = section.getUpStation();
            final Station downStation = section.getDownStation();

            graph.addVertex(upStation);
            graph.addVertex(downStation);
            graph.setEdgeWeight(graph.addEdge(upStation, downStation), section.getDistance());
            graph.setEdgeWeight(graph.addEdge(downStation, upStation), section.getDistance());
        }
    }

    public GraphPath<Station, DefaultWeightedEdge> findShortestPath(final Station source, final Station target) {
        return dijkstraShortestPath.getPath(source, target);
    }
}

to-be:
public interface PathCalculator {
    GraphPath<Station, DefaultWeightedEdge> findShortestPath(final Station source, final Station target);
}

wooteco.subway.infrastructure:
public class PathCalculatorDijkstra implements PathCalculator {

    private final DijkstraShortestPath<Station, DefaultWeightedEdge> dijkstraShortestPath;

    public PathCalculatorDijkstra(final List<Line> lines) {
        final WeightedMultigraph<Station, DefaultWeightedEdge> graph = new WeightedMultigraph<>(
                DefaultWeightedEdge.class);
        for (Line line : lines) {
            final List<Section> sections = line.getSections();
            addSectionInGraph(graph, sections);
        }
        dijkstraShortestPath = new DijkstraShortestPath<>(graph);
    }

    private void addSectionInGraph(final WeightedMultigraph<Station, DefaultWeightedEdge> graph,
                                   final List<Section> sections) {
        for (Section section : sections) {
            final Station upStation = section.getUpStation();
            final Station downStation = section.getDownStation();

            graph.addVertex(upStation);
            graph.addVertex(downStation);
            graph.setEdgeWeight(graph.addEdge(upStation, downStation), section.getDistance());
            graph.setEdgeWeight(graph.addEdge(downStation, upStation), section.getDistance());
        }
    }

    @Override
    public GraphPath<Station, DefaultWeightedEdge> findShortestPath(final Station source, final Station target) {
        return dijkstraShortestPath.getPath(source, target);
    }
}

skullkim committed:
refactor(PathCalculator.java): 라이브러리를 도메인 에서 제거

도메인에 존재하던 라이브러리를 도메인 격리를 위해 분리

BREKAING CHANGE: 경로 조회 관련 클래스를 인스턴스로 변경
as-is:
public interface PathCalculator {
    GraphPath<Station, DefaultWeightedEdge> findShortestPath(final Station source, final Station target);
}

@Service
public class PathService {

    ...

    public PathResponse findShortestPath(final PathRequest request) {
        ...

        final PathCalculator pathCalculator = new PathCalculatorDijkstra(lines);
        final GraphPath<Station, DefaultWeightedEdge> path = pathCalculator.findShortestPath(source, target);
        final List<Station> stations = path.getVertexList();
        final int distance = (int) path.getWeight();
        ...
        return new PathResponse(stations, distance, fare);
    }
}

to-be:
subway/domain/path/Path.java:
public interface Path {
    List<Station> getVertexList();
    int getWeight();
}

subway/infrastructure/SubwayPath.java:
public class SubwayPath implements Path {

    private final GraphPath<Station, DefaultWeightedEdge> path;

    public SubwayPath(GraphPath<Station, DefaultWeightedEdge> path) {
        this.path = path;
    }

    @Override
    public List<Station> getVertexList() {
        return path.getVertexList();
    }

    @Override
    public int getWeight() {
        return (int) path.getWeight();
    }
}

/subway/domain/path/PathCalculator.java:
public interface PathCalculator {
    Path findShortestPath(final Station source, final Station target);
}

subway/infrastructure/PathCalculatorDijkstra.java:
public class PathCalculatorDijkstra implements PathCalculator {

    ...

    @Override
    public Path findShortestPath(final Station source, final Station target) {
        return new SubwayPath(dijkstraShortestPath.getPath(source, target));
    }
}

@Service
public class PathService {

    ...

    public PathResponse findShortestPath(final PathRequest request) {
        ...

        final PathCalculator pathCalculator = new PathCalculatorDijkstra(lines);
        final Path path = pathCalculator.findShortestPath(source, target);
        final List<Station> stations = path.getVertexList();
        final int distance = path.getWeight();

        ...
        return new PathResponse(stations, distance, fare);
    }
}

skullkim committed:
refactor(Fare.java): 책임 분리

경로에서 특정 라인 존재여부를 확인하는 기능을 라인 컬렉션으로 이동

BREAKING CHANGE: calculateExtraLineFare 메서드를 Lines로 이동
as-is:
public class Fare {

    ...

     public int calculate(final int distance, final List<Station> stations, final List<Line> lines) {
        final int extraLineFare = calculateExtraLineFare(stations, lines);
        ...
    }

    private int calculateExtraLineFare(final List<Station> stations, final List<Line> lines) {
        return lines.stream()
                .filter(line -> doesLineContainStation(line, stations))
                .mapToInt(Line::getExtraFare)
                .max()
                .orElse(DEFAULT_EXTRA_LINE_FARE);
    }

    private boolean doesLineContainStation(final Line line, final List<Station> stations) {
        return stations.stream().anyMatch(line::containStation);
    }

    ...
}

to-be:

public class Lines {

    private static final int DEFAULT_EXTRA_LINE_FARE = 0;

    private final List<Line> values;

    public Lines(List<Line> values) {
        this.values = values;
    }

    public int calculateExtraLineFare(final List<Station> stations) {
        return values.stream()
                .filter(line -> doesLineContainStation(line, stations))
                .mapToInt(Line::getExtraFare)
                .max()
                .orElse(DEFAULT_EXTRA_LINE_FARE);
    }

    private boolean doesLineContainStation(final Line line, final List<Station> stations) {
        return stations.stream().anyMatch(line::containStation);
    }
}

public class Fare {
    ...

    private final Lines lines;
    ...

    public Fare(final Lines lines, final AgeDiscountPolicy ageDiscountPolicy) {
        this.lines = lines;
        ...
    }

    public int calculate(final int distance, final List<Station> stations) {
        final int extraLineFare = lines.calculateExtraLineFare(stations);
        ...
    }

    ...
}

skullkim committed:
chore(Fare.java): 필요 없는 코드 삭제

skullkim committed:
chore(domain/): 하위 패키지 생성

명확한 구조를 위해 하위 패키지 생성

wbluke commented on src/main/java/wooteco/subway/service/PathService.java:
```diff
@@ -24,16 +28,18 @@ public PathService(final StationRepository stationRepository, final LineReposito
-    public PathResponse findShortestPath(final Long sourceId, final Long targetId, final int age) {
-        final Station source = stationRepository.findById(sourceId);
-        final Station target = stationRepository.findById(targetId);
+    public PathResponse findShortestPath(final PathRequest request) {
+        final PathDto pathDto = PathDto.from(request);
```
계층별 DTO를 따로 만들어보라고 했던 이유는 사실 계층 간 의존 방향에 대한 내용을 의도한 것이었는데요 ㅎㅎ
기능 단위를 보장하는 Service 계층에서는 (여러 개일 수도 있는) 외부 계층인 Web 계층의 DTO를 모르는 상태면 좋을 것 같아요.
```java
service.findShortestPath(request.toServiceRequest());
```
그래서 Controller쪽에서 DTO 변환을 수행하고, DTO 끼리도 `PathDto.from(request);` 보다 데이터를 가지고 있는 Web쪽 DTO에서 위와 같은 형태로 변환하는 것이 좋을 것 같아요.
의존 방향이 명확하게 나눠지면 추후에 모듈을 분리하거나 할 때도 깔끔하게 분리가 될거에요 :)

wbluke commented on src/main/java/wooteco/subway/ui/PathController.java:
```diff
@@ -2,13 +2,15 @@
 import wooteco.subway.service.PathService;
 
 @RestController
+@Validated
```
`@Validated`와 `@Valid`를 같이 쓰면 아마 2중으로 검증이 될거에요. (각자 동작하는 위치가 달라서)
당장 문제는 없을테지만 잘 알고 쓰는 것이 좋습니다 :)

wbluke commented on src/test/java/wooteco/subway/ui/Converter.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.test.web.servlet.MvcResult;
+import wooteco.subway.dto.response.ErrorResponse;
+
+public class Converter {
```
테스트용 컨버터이긴 하지만 네이밍을 상세하게 지으면 좋을 것 같네요 :)

wbluke commented on src/test/java/wooteco/subway/ui/Converter.java:
```diff
@@ -0,0 +1,18 @@
+
+public class Converter {
+
+    public static String ConvertToStringFrom(final MvcResult result, final ObjectMapper objectMapper)
```
메서드 네이밍!

wbluke reviewed (approved):
리팩토링 잘 해주셨네요 👍 
참고할 코멘트 조금 남겼으니 확인해주시고 경로 미션 고생 많으셨어요~! 🙂
